### PR TITLE
New extensions API, supporting maturity levels

### PR DESCRIPTION
**Pull request by goldfirere**
Recommended reviewer: @ncik-roberts (because @antalsz is out).

### CODE REVIEW

File: ocaml/parsing/jane_syntax_parsing.ml [hunk 1]
```diff
@@ -98,11 +98,11 @@ module Feature : sig
 
   val is_erasable : t -> bool
 end = struct
-  type t = Language_extension of Language_extension.t
+  type t = Language_extension : 'a Language_extension.t -> t
          | Builtin
 
   type error =
-    | Disabled_extension of Language_extension.t
+    | Disabled_extension of Language_extension.Exist.t
```
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax_parsing.ml:105]: `Language_extension` above is a GADT constructor; `Disabled_extension` is a non-GADT constructor taking a packed extension. Can we do the same thing for both? I don't much care which. Feel free to resolve upon addressing.

File: ocaml/utils/language_extension.ml [hunk 2]
```diff
@@ -201,3 +227,75 @@ let disallow_extensions () =
   (* The strictest option, so no internal checks needed *)
   extensions := [];
   universe   := Universe.No_extensions
+
+(* converts full extension names, like "layouts_alpha" to a pair of
+   an extension and its setting *)
+let pair_of_string extn_name : extn_pair option =
```
Comment (ncik-roberts) [inline ocaml/utils/language_extension.ml:233]: This should be next to `to_string` so it's easier to see they correspond.

File: ocaml/typing/typetexp.mli [hunk 3]
```diff
@@ -95,7 +95,7 @@ type error =
   | Method_mismatch of string * type_expr * type_expr
   | Opened_object of Path.t option
   | Not_an_object of type_expr
-  | Unsupported_extension of Language_extension.t
+  | Unsupported_extension : 'a Language_extension.t -> error
```
Comment (ncik-roberts) [inline ocaml/typing/typetexp.mli:98]:
```suggestion
  | Unsupported_extension : _ Language_extension.t -> error
```

File: ocaml/typing/typetexp.ml [hunk 4]
```diff
@@ -54,7 +54,7 @@ type error =
   | Method_mismatch of string * type_expr * type_expr
   | Opened_object of Path.t option
   | Not_an_object of type_expr
-  | Unsupported_extension of Language_extension.t
+  | Unsupported_extension : 'a Language_extension.t -> error
```
Comment (ncik-roberts) [inline ocaml/typing/typetexp.ml:57]:
```suggestion
  | Unsupported_extension : _ Language_extension.t -> error
```

File: ocaml/parsing/jane_syntax_parsing.ml [hunk 5]
```diff
@@ -83,11 +83,11 @@ open Parsetree
 
 module Feature : sig
   type t =
-    | Language_extension of Language_extension.t
+    | Language_extension : 'a Language_extension.t -> t
```
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax_parsing.ml:86]:
```suggestion
    | Language_extension : _ Language_extension.t -> t
```

Here and throughout, I've made this suggestion to be more suggestive of a type variable that doesn't corefer.

File: ocaml/utils/language_extension.ml [hunk 6]
```diff
@@ -201,3 +227,75 @@ let disallow_extensions () =
...
+  let to_string : t -> string = function
+    | Pack extn -> to_string extn
+
+  let is_enabled : t -> bool = function
+    | Pack extn -> is_enabled extn
+
+  let is_erasable : t -> bool = function
+    | Pack extn -> is_erasable extn
+
+  let all : t list =
```
Comment (ncik-roberts) [inline ocaml/utils/language_extension.ml:281]: It seems easy to forget to add something to `all`.

A low-tech way of making this less likely would be to move the definition of `all` up higher in the file, perhaps closer to one of the exhaustive matches like `equal_t` or `to_string`. (It seems fine to me to e.g. define `Pack` and `all` at the top-level of the ml file and then re-export it, and `all`, from `Exist`.)

File: ocaml/utils/language_extension.mli [hunk 7]
```diff
@@ -1,53 +1,81 @@
...
-val to_string : t -> string
-val of_string : string -> t option
-val of_string_exn : string -> t
+val to_string : 'a t -> string
+val of_string : string -> Exist.t option
+
+(** Enable and disable according to command-line strings; these raise
+    an exception if the input string is invalid. *)
+val enable_of_string_exn : string -> unit
+val disable_of_string_exn : string -> unit
```
Comment (ncik-roberts) [inline ocaml/utils/language_extension.mli:57]: Is it expected that enabling `layouts` but disabling `layouts_alpha` will result in `layouts` being disabled?

**Review by ncik-roberts — APPROVED** [review]
Looks good to me. I just have minor comments/questions, so I'm approving.

File: ocaml/utils/language_extension.mli (hunk 7 above)
Comment (goldfirere) [reply ocaml/utils/language_extension.mli:58]: Yes. I ran this change in semantics by Chris and Carl.

File: ocaml/utils/language_extension.ml [hunk 8]
```diff
@@ -111,42 +151,95 @@ let is_erasable : type a. a t -> bool = function
...
     | Only_erasable -> is_erasable ext
     | Any           -> true
+
+  (* are _all_ extensions allowed? *)
+  let all_allowed () = match !universe with
+    | Any -> true
+    | No_extensions | Only_erasable -> false
+
+  (* Why the [()]? The value restriction. *)
+  let fail () = Format.ksprintf (fun str -> raise (Arg.Bad str))
```
Comment (ncik-roberts) [inline ocaml/utils/language_extension.ml]: I've pushed a commit to refactor this in the style of `Base.failwithf`, which has the advantage that you're less likely to pass more arguments than format specifiers. (Feel free to resolve this.)

**Review by ncik-roberts — APPROVED** [review]
LGTM.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**